### ISSUE TITLE
Provide support for Jiff date/time types

### ISSUE THREAD

**Issue by sgoll** [graphql-rust/juniper#1270]
## Description

This is a proposal to add first-class support for the data types from [Jiff](https://crates.io/crates/jiff). First-class support is necessary, and an external crate would be much more inconvenient to use, because of Rust's orphan rule as outlined in https://github.com/graphql-rust/juniper/issues/87#issuecomment-325918618.

This should cover the following date/time types:

### Timestamps

- [`Timestamp`](https://docs.rs/jiff/0.1.4/jiff/struct.Timestamp.html), serialized as RFC 3339 (ISO 8601) date/time with UTC time zone, GraphQL scalar [`DateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/date-time), e.g. `2024-06-19T19:22:45Z`
- ~[`Zoned`](https://docs.rs/jiff/0.1.4/jiff/struct.Zoned.html), serialized was RFC 3339 (ISO 8601) date/time with underlying time zone, GraphQL scalar `DateTime`, e.g. `2024-06-19T15:22:00-04:00`[^1]~

[^1]: ~Alternatively, this could include the time zone specifier and use RFC ~8536~ 9557 instead, i.e. `2024-07-04T08:39:00-04:00[America/New_York]`. This would probably be harder to consume, though, because that format seems not as widely used. It is also not a valid input for GraphQL scalar `DateTime`.~

### Wall-clocked

- [`Date`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.Date.html), serialized as date, without time zone, GraphQL scalar [`LocalDate`](https://the-guild.dev/graphql/scalars/docs/scalars/local-date), e.g. `2024-06-19`
- [`Time`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.Time.html), serialized as time, without time zone, GraphQL scalar [`LocalTime`](https://the-guild.dev/graphql/scalars/docs/scalars/local-time), e.g. `15:22:45`
- [`DateTime`](https://docs.rs/jiff/0.1.4/jiff/civil/struct.DateTime.html), serialized as date/time, without time zone, GraphQL scalar [`LocalDateTime`](https://the-guild.dev/graphql/scalars/docs/scalars/local-date-time), e.g. `2024-06-19T15:22:45`

### Durations

- [`Span`](https://docs.rs/jiff/0.1.4/jiff/struct.Span.html), serialized as ISO 8601 duration, GraphQL scalar [`Duration`](https://the-guild.dev/graphql/scalars/docs/scalars/duration), e.g. `P5dT8h1m`

The resulting serialization would be identical to the one for the [`chrono`](https://crates.io/crates/chrono) and [`time`](https://crates.io/crates/time) crates, with the difference of using GraphQL scalar `LocalDate` instead of [`Date`](https://the-guild.dev/graphql/scalars/docs/scalars/date).[^2]

[^2]: `LocalDate` seems to be the more appropriate type. I am not sure why `chrono` and `time` integration chose to use `Date` instead.

I am aware that Jiff is a very new library but I think it has great potential in the ecosystem. Having dealt with date/time intricacies in several projects, I regard it as one of the top-contenders for following or eventually even superseding `chrono`/`time`.

~I am willing to provide a PR if this proposal is accepted.~ See #1271 for a PR that implements these features.

**Comment by BurntSushi** [comment]
Jiff author here. Great to see this! Happy to answer questions. :-)

For `Zoned`, in order to deserialize into it, a time zone ID like `[America/New_York]` is required. This is from [RFC 9557](https://datatracker.ietf.org/doc/rfc9557/) (not RFC 8536) and extends RFC 3339. You are absolutely correct that it does not yet enjoy broad support. Temporal will support it (but that isn't in browsers yet) and `java.time` supports it. I'm not sure what else does, but hopefully more will support it soon.

With that said, if your datetimes do not have a time zone ID in them, and GraphQL cannot otherwise store them, then probably the right thing to do is to _not_ offer `Zoned` support directly. Either that, or error if the `Zoned` contains a time zone with a non-None [`TimeZone::iana_name`](https://docs.rs/jiff/latest/jiff/tz/struct.TimeZone.html#method.iana_name), since otherwise serializing and deserializing will be lossy.

The idea with a `Zoned` is that if you have a `Zoned` in hand, then you have an assertion about what the local time is in relationship to UTC.

Everything else looks correct to me, including the mapping from Jiff's civil date types to GraphQL's local data types. (With the caveat that I don't know too much about GraphQL. But the docs you linked all seem like they support the mapping you chose here.)

**Comment by BurntSushi** [comment]
1. I think this should be fine. If GraphQL outright rejects datetimes with precision better than milliseconds, then I think you'll just need to strip anything less than milliseconds from the datetime (possibly via rounding). Otherwise, we can teach `jiff::fmt::temporal::DateTimePrinter` a new knob for controlling output precision.
2. Yeah that's a tough one, but I think that fits into my suggestion. That is, you return an error when trying to serialize a `Zoned` with an IANA time zone. But you specifically allow serializing a `Zoned` when its `TimeZone::iana_name` is none. This permits users to serialize a fixed offset `Zoned`. And if they have an IANA time zone name, they'll need to do an explicit conversion, e.g., `zdt.with_time_zone(TimeZone::fixed(zdt.offset()))`. But the point is that it's explicit and they aren't subjecting themselves to a subtle footgun. And this is a _very_ subtle footgun, because a fixed offset will "just work" in a lot of cases.

**Comment by BurntSushi** [comment]
> In my current draft, I am rounding [manually](https://github.com/sgoll/juniper/blob/1674c876df58568fb2b8aca682d55cdb2d5b3f20/juniper/src/integrations/jiff.rs#L127). The need to handle errors is unfortunate, but since these can happen only for configuration errors (which we can rule out), there should be no problems at runtime.

Yeah that works. You could also drop the rounding if you're okay with truncation, which is what the `%.3f` will do.

> Thus, I tend to agree with your initial assessment https://github.com/graphql-rust/juniper/issues/1270#issuecomment-2265865082 and would probably indeed not offer serialization for Zoned—which is unfortunate, but it still seems better than having to pick between two footguns 😐

Holding off on a `Zoned` impl is fine I think. But to push back a little, presuming serialization returns an error, it doesn't seem like it'd be a huge footgun since the caller will get an error if they try to do something wrong. Then they can fix it. It's at least better than the footgun of silently dropping information.

**Comment by BurntSushi** [comment]
Gotya. Yeah I'm not especially familiar with GraphQL, but I think what you're saying there makes sense.

I think it would be good to leave out `Zoned` for now. Not necessarily because we ought to wait for better GraphQL support, but also just to see what practice demands. If the practical advantages of supporting `Zoned` end up being large enough to outweigh the footgun of dropping the time zone, then it might make sense to add it at that point.

**Comment by tyranron** [comment]
@sgoll 

> 1. Leave out `Zoned` serialization for now, only serialize `Timestamp` into GraphQL scalar `DateTime`. Then, when some new GraphQL scalar _ZonedDateTime_[2](#user-content-fn-2-a16f334093af5d881e7606f038e33448) has been added, we serialize `Zoned` into it.

The fact, that https://graphql-scalars.dev/docs/scalars/ collection doesn't provide such a scalar is not a blocker as for me. We just can introduce the `ZonedDateTime` scalar by ourselves and don't bother, as we had done with the `LocalDateTime` scalar before it appeared there, and as we do with some scalars absent there even now (like `BigDecimal`).
